addons/account/models/account_account.py

```diff
@@ -432,41 +628,68 @@ def code_is_available(new_code):
 
         raise UserError(_('Cannot generate an unused account code.'))
 
+    @api.depends_context('allowed_company_ids')
     def _compute_current_balance(self):
-        balances = {
-            account.id: balance
-            for account, balance in self.env['account.move.line']._read_group(
-                domain=[('account_id', 'in', self.ids), ('parent_state', '=', 'posted')],
-                groupby=['account_id'],
-                aggregates=['balance:sum'],
-            )
+        balances = self.env['account.move.line']._read_group(
+            domain=[
+                ('account_id', 'in', self.ids),
+                ('parent_state', '=', 'posted'),
+                *self.env['account.move.line']._check_company_domain(self.env.company),
+            ],
+            groupby=['account_id', 'currency_id'],
+            aggregates=['balance:sum'],
+        )
+        line_currencies = self.env['res.currency'].union(*(currency for _account, currency, _balance in balances))
+        currency_rates = (line_currencies | self.env.company.currency_id)._get_rates(self.env.company, fields.Date.context_today(self))
+        user_company_rate = currency_rates[self.env.company.currency_id.id]
+        conversion_rates = {
+            currency_id: user_company_rate / rate
+            for currency_id, rate in currency_rates.items()
+        }
+        balance_by_account = {
+            account: self.env.company.currency_id.round(sum(
+                balance * conversion_rates[currency.id]
+                for _account, currency, balance in balances_grouped
+            ))
+            for account, balances_grouped in groupby(balances, lambda x: x[0])
         }
         for record in self:
-            record.current_balance = balances.get(record.id, 0)
+            record.current_balance = balance_by_account.get(record, 0.0)
 
+    @api.depends_context('allowed_company_ids')
     def _compute_related_taxes_amount(self):
         for record in self:
             record.related_taxes_amount = self.env['account.tax'].search_count([
                 ('repartition_line_ids.account_id', '=', record.id),
+                *self.env['account.tax']._check_company_domain(self.env.companies),
```

Comment:
Is this doing anything?
Isn't it already part of the record rules?

Reply:
ummm yes.